nixos/modules/misc/assertions.nix

```diff
@@ -1,34 +1,155 @@
 { lib, ... }:
+{
+  options = {
+    assertions = lib.mkOption {
+      type = let
+        assertionItemType = lib.types.submodule {
+          options = {
+            enable = lib.mkOption {
+              description = ''
+                Whether to enable this assertion.
 
-with lib;
+                This option is mostly useful for users, in order to forcefully disable assertions they believe to be
+                erroneous while waiting for someone to fix the assertion upstream.
+              '';
+              type = lib.types.bool;
+              default = true;
+              example = false;
+            };
 
-{
+            assertion = lib.mkOption {
+              description = "Condition to be asserted. If this is `false`, the evaluation will throw the error";
+              type = lib.types.bool;
+              example = false;
+            };
 
-  options = {
+            message = lib.mkOption {
+              description = "The contents of the error message that should be shown upon triggering a false assertion";
+              type = lib.types.nonEmptyStr;
+              example = "This is an example error message";
+            };
+          };
+        };
 
-    assertions = mkOption {
-      type = types.listOf types.unspecified;
+        # This might be replaced when tagged submodules or similar are available,
+        # see https://github.com/NixOS/nixpkgs/pull/254790
+        checkedAssertionItemType = let
+          check = x: x ? assertion && x ? message;
+        in lib.types.addCheck assertionItemType check;
+
+        nestedAssertionAttrsType = with lib.types; let
+          nestedAssertionItemType = oneOf [
+            checkedAssertionItemType
+            (attrsOf nestedAssertionItemType)
+          ];
+        in nestedAssertionItemType // {
+          description = "nested attrs of (${assertionItemType.description})";
+        };
+
+        # Backwards compatibility for assertions that are still written as attrs inside a list.
+        # The attribute name will be set to the sha256 sum of the assertion message, e.g. `assertions.legacy."<sha256>" = { ... }`
+        coercedAssertionAttrs = let
+          coerce = xs: {
+            legacy = lib.listToAttrs (map (assertion: lib.nameValuePair (builtins.hashString "sha256" assertion.message) assertion) xs);
+          };
```

Comment:
This might be more easily described by https://github.com/NixOS/nixpkgs/pull/270127